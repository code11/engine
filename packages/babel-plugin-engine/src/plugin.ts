import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
