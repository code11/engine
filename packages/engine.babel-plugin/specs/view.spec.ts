import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "../src";

const config = require("./babelConfig.json");

// Add tests regarding let foo: producer = () => {}, bar: producer = () => {}

// Test invalid syntax with const foo: view = 123

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  pluginOptions: {
    viewLibrary: "engineViewLibrary",
  },
  pluginName: "babel-plugin-engine",
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should compile a view": {
      code: `
      const foo: view = ({ foo = observe.foo }) => {}
      `,
      snapshot: true,
    },
    "should add the import statement only once": {
      code: `
      const foo: view = ({ foo = observe.foo }) => { }
      const bar: view = ({ foo = observe.foo }) => { }
      `,
      snapshot: true,
    },
    "should throw if not given an arrow function": {
      code: `const foo: view = 123`,
      error: true,
    },
    "should throw if the function param is not an object": {
      code: `const foo: view = (foo) => {}`,
      error: true,
    },
    "should throw if the variable declaration is an object pattern": {
      code: `const { foo }: view = () => {}`,
      error: true,
    },
    "should support empty arguments": {
      code: `
        const foo: view = () => {}
      `,
      snapshot: true,
    },
  },
});
