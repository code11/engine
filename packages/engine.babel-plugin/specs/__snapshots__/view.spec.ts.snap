// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-engine should add the import statement only once: should add the import statement only once 1`] = `

const foo: view = ({ foo = observe.foo }) => { }
const bar: view = ({ foo = observe.foo }) => { }

      ↓ ↓ ↓ ↓ ↓ ↓

import { view as viewEngine } from "engineViewLibrary";
const foo = viewEngine({
  props: {
    type: "STRUCT",
    value: {
      foo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
    },
    meta: {},
  },
  meta: {},
  fn: ({ foo: foo }) => {},
});
const bar = viewEngine({
  props: {
    type: "STRUCT",
    value: {
      foo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
    },
    meta: {},
  },
  meta: {},
  fn: ({ foo: foo }) => {},
});


`;

exports[`babel-plugin-engine should compile a view: should compile a view 1`] = `

const foo: view = ({ foo = observe.foo }) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

import { view as viewEngine } from "engineViewLibrary";
const foo = viewEngine({
  props: {
    type: "STRUCT",
    value: {
      foo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
        ],
      },
    },
    meta: {},
  },
  meta: {},
  fn: ({ foo: foo }) => {},
});


`;

exports[`babel-plugin-engine should support empty arguments: should support empty arguments 1`] = `

const foo: view = () => {}

      ↓ ↓ ↓ ↓ ↓ ↓

import { view as viewEngine } from "engineViewLibrary";
const foo = viewEngine({
  props: {
    type: "STRUCT",
    value: {},
    meta: {},
  },
  meta: {},
  fn: ({}) => {},
});


`;
